indexes: Read the locator's top block during init, allow interaction with reindex-chainstate

mzumsande opened the pull request:
This makes two improvements to the index init phase:

**1) Prevent index corruption in case a reorg happens when the index was switched off**:
This is done by reading in the top block stored in the locator instead of looking for a fork point already in `BaseIndex::Init()`.
Before, we'd just go back to the fork point by calling `FindForkInGlobalIndex()`, which would have corrupted the coinstatsindex because its saved muhash needs to be reverted step by step by un-applying all blocks in between, which wasn't done before. This is now being done a bit later in  `ThreadSync()`, which has existing logic to call the custom `Rewind()` method when going back along the chain to the forking point (thanks ryanofsky for pointing this out to me!).

**2) Allow using the `-reindex-chainstate` option without needing to disabling indexes**:
With `BaseIndex::Init()` not calling `FindForkInGlobalIndex()` anymore, we can allow `reindex-chainstate` with active indexes. `reindex-chainstate` deletes the chain and rebuilds it later in `ThreadImport`, so there is no chain available during `BaseIndex::Init()`, which would lead to problems (see #24789).
But now we'll only need the chain a bit later in `BaseIndex::ThreadSync`, which will wait for the reindex-chainstate in `ThreadImport` to finish and will continue syncing after that.

mzumsande committed:
index: Use first block from locator instead of looking for fork point

The index sync code has logic to go back the chain to the forking point, while
also updating index-specific state, which is necessary to prevent
possible corruption of the coinstatsindex.

Also add a test for this (a reorg happens while the index is deactivated)
that would not pass before this change.

mzumsande committed:
index: Enable reindex-chainstate with active indexes

This is achieved by letting the index sync thread wait until
reindex-chainstate is finished.

This also disables the pruning check when reindexing the chainstate (which is
incompatible with prune mode) because there would be no chain at this point
in init.

mzumsande commented:
One thing I'm unsure about is that part 1) of this PR  will now make it impossible to go back if we don't know the top block of the locator for some reason - although I don't know how this would be possible because I don't know of a process that would prune stale blocks from the Block Index (except the contrib/linearize script).

The old code could recover an index with a no-longer existing best block from that (at the cost of corrupting the coinstatsindex), but I wonder if it would be good to still call `FindForkInGlobalIndex()`, maybe as a fallback, for the other indexes.

jonatack commented:
Concept ACK on reading the changes

ryanofsky commented on src/node/blockstorage.cpp:
```diff
@@ -885,6 +886,7 @@ void ThreadImport(ChainstateManager& chainman, std::vector<fs::path> vImportFile
                 StartShutdown();
                 return;
             }
+            fReindexChainState = false;
```
In commit "node: add fReindexChainState flag to node" (8898261ac03945e559c98a0abcb9455942ab4775)

I understand what this are doing with the new `fReindexChainState` global is similar to what happens with the existing `fReindex` global, but I think what happens with the `fReindex` global is not ideal. I think it's unnecessarily fragile how the value starts off false, then switches to true, then switches back to false again, and think a one way latch would be better. I think it's bad the the variable doesn't have any straightforward meaning but is some combination "was reindexing requested?" and "is reindexing in progress?". Also in this commit, `fReindexChainState` is different local variables subtly replaced by a global so you can't tell just looking at a diff whether all the existing `fReindexChainState` references were updated correctly.

I'd suggest leaving `fReindexChainState` alone here, and adding a simpler `atomic_bool g_indexes_ready_to_sync = false` global variable, and setting it to true here, and in the normal code path. I think this would make the diff and the overall initialization sequence simpler and easier to understand.

mzumsande replied:
That makes sense. I dropped the commit that made `fReindexChainState` a global and added `g_indexes_ready_to_sync` as suggested.

ryanofsky commented on src/index/base.cpp:
```diff
@@ -76,7 +78,7 @@ bool BaseIndex::Init()
     }
     m_synced = m_best_block_index.load() == active_chain.Tip();
-    if (!m_synced) {
+    if (!m_synced && fPruneMode) {
```
In commit "index: Enable reindex-chainstate with active indexes" (1e598cfb2b3d0e56b1fb301bcf6870f42453c7cc)

This is pretty opaque and could use a comment. I'm also not sure it is better to be looking at `fPruneMode` here, than just checking whether reindexing is happening, when it sounds like reindexing is the real problem, and checking for `fPruneMode` is just a proxy for avoiding the real source of the problem?

Also it is not obvious to me that just because pruning isn't currently enabled doesn't mean pruning wasn't previously enabled and there couldn't be missing blocks worth checking for here.

mzumsande replied:
> Also it is not obvious to me that just because pruning isn't currently enabled doesn't mean pruning wasn't previously enabled and there couldn't be missing blocks worth checking for here.

I think that this shouldn't be possible: We check in Init when loading the chainstate (`chainman.LoadBlockIndex()`) that we either have all the blocks from genesis, or allow for pruning - otherwise, we abort with an InitError and never get to the point where the indexes are started.

I replaced the check with the indirect `g_indexes_ready_to_sync` as suggested and added a comment. This way, the pruning check will continue to be executed regardless of pruning status (unless -reindex is specified).

What I don't like is that this would throw a confusing InitError message if it somehow failed on a non-pruning node ("block of the index goes beyond pruned data"). Maybe the check should assert instead if `fPrune==false`?

ryanofsky reviewed (approved):
Light code review ACK 1e598cfb2b3d0e56b1fb301bcf6870f42453c7cc. First commit looks good. I had some questions and suggestions about some things in the second and third commits, but all the code looked ok and seemed like it should work

Crypt-iQ commented on src/index/base.cpp:
```diff
@@ -67,10 +69,16 @@ bool BaseIndex::Init()
-        SetBestBlockIndex(m_chainstate->FindForkInGlobalIndex(locator));
+        // Setting the best block to the locator's top block. If it is not part of the
+        // best chain, we will rewind to the fork point during index sync
+        const CBlockIndex* locator_index{m_chainstate->m_blockman.LookupBlockIndex(locator.vHave.front())};
```
Just leaving as a note, but I think this allows the `if (!active_chain.Contains(block_to_test)) {` case to be hit. Previously `FindForkInGlobalIndex` would return a `CBlockIndex` only in the current chain and so `active_chain.Contains` would always be true.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/25193#discussion_r904094952

In commit "index: Use first block from locator instead of looking for fork point" (20bd2219894e53ab39d2b11edf9a7f3c7bda489b)

Yes, this seems to be the case. This is a good indication the original author or this code expecting was expecting to FindForkInGlobalIndex to just return a block based on the locator, not necessarily a block on the current chain. So new code is probably closer to intent of original code.

Bat3an replied:
Thank you your a angle

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/25193#discussion_r907572547

> Without this patch, is it possible that the old code goes back to genesis in this rare, kind of contrived scenario?:

Maybe I need to think about this more, but I don't think it can go back to genesis. It can just go back to the last common block before the reorg. Also, the problem which this patch fixes isn't going backwards in general, but going backwards without rewinding. The problem with the `FindForkInGlobalIndex` is that it goes backwards without making needed `Rewind()` call to update indexes

Crypt-iQ replied:
From my reading, if `FindForkInGlobalIndex` doesn't find `pindex` in the chain or doesn't find `Tip()` as an ancestor of `pindex`, it returns genesis. But yeah a bit orthogonal to the issue at hand

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/25193#discussion_r907779757

> From my reading, if `FindForkInGlobalIndex` doesn't find `pindex` in the chain or doesn't find `Tip()` as an ancestor of `pindex`, it returns genesis. But yeah a bit orthogonal to the issue at hand

Wow, you are right. The [`FindForkInGlobalIndex`](https://github.com/bitcoin/bitcoin/blob/2111f32f2a6998531871e7792b5208992868ba7f/src/validation.cpp#L131-L149) behavior is much different than the [`FindFork`](https://github.com/bitcoin/bitcoin/blob/2111f32f2a6998531871e7792b5208992868ba7f/src/chain.cpp#L54-L63) behavior. I assumed `FindForkInGlobalIndex` would try to find the last common ancestor between the locator block and the chain like `FindFork` does, but it doesn't even try to do this. Instead it will literally only return the exact locator block, or the chain tip, or the genesis block. It doesn't make any sense to me why the function would be implemented this way, but I guess it's good that this PR removes one usage of it...

mzumsande commented on src/index/base.cpp:
```diff
@@ -67,7 +67,13 @@ bool BaseIndex::Init()
-        SetBestBlockIndex(m_chainstate->FindForkInGlobalIndex(locator));
+        // Setting the best block to the locator's top block. If it is not part of the
+        // best chain, we will rewind to the fork point during index sync
+        const CBlockIndex* locator_index{m_chainstate->m_blockman.LookupBlockIndex(locator.vHave.front())};
```
Changed to `vHave.at(0)`

mzumsande commented:
[1e598cf ](https://github.com/bitcoin/bitcoin/commit/1e598cfb2b3d0e56b1fb301bcf6870f42453c7cc)to [9bcd930](https://github.com/bitcoin/bitcoin/commit/9bcd930203525b871cdcc7868141e7bbc6b28f45):
addressed feedback by @ryanofsky - thanks!

ryanofsky commented on src/init.cpp:
```diff
@@ -1563,6 +1551,9 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     // ********************************************************* Step 8: start indexers
+
+    // If reindex-chainstate was specified, delay syncing indexes until ThreadImport has reindexed the chain
+    g_indexes_ready_to_sync = !fReindexChainState;
```
In commit "index: Enable reindex-chainstate with active indexes" (91947e4387d98290d39c9201f85a9aefa235a4d1)

This is a pedantic, academic suggestion, but should consider writing this as:

```c++
if (!fReindexChainState) g_indexes_ready_to_sync = true;
```

To make this part of the init function idempotent, and also make it clearer that index code only needs to handle `g_indexes_ready_to_sync` transitioning from false to true, never the other way around.

mzumsande replied:
Done with the latest push.

ryanofsky reviewed (approved):
Code review ACK 91947e4387d98290d39c9201f85a9aefa235a4d1. In first commit vector `front()` call is replaced by `at(0)` call to avoid undefined behavior in case null locator is loaded. In second commit `fReindexChainState` variable is replaced by simpler `g_indexes_ready_to_sync` that only changes from false to true

mzumsande commented:
[91947e4 ](https://github.com/bitcoin/bitcoin/commit/91947e4387d98290d39c9201f85a9aefa235a4d1)to [702f481](https://github.com/bitcoin/bitcoin/commit/702f481f7e97ebfd438f27464de3a2d4d77497bc): 
rebased and addressed comment by @ryanofsky

ryanofsky reviewed (approved):
Code review ACK 702f481f7e97ebfd438f27464de3a2d4d77497bc. Only changes since last review were rebasing and making suggested code clarification (thanks!)

mzumsande commented:
[702f481 ](https://github.com/bitcoin/bitcoin/commit/702f481f7e97ebfd438f27464de3a2d4d77497bc)to [6a85526](https://github.com/bitcoin/bitcoin/commit/6a855268964152f920baa98b8e71150604043448):
rebased due to conflict with #25487

mzumsande commented:
[6a85526 ](https://github.com/bitcoin/bitcoin/commit/6a855268964152f920baa98b8e71150604043448)to [53ff637](https://github.com/bitcoin/bitcoin/commit/53ff63799404e6828779c5e4f321acf256fe9018):
rebased due to conflict with #25308

ryanofsky reviewed (approved):
Code review ACK 53ff63799404e6828779c5e4f321acf256fe9018. Only changes since last review were rebasing.

I do think it would be good if PR description described benefits of these changes, which are (1) avoiding possible coinstatsindex corruption in the case of a reorg and (2) allowing the `-reindex-chainstate` option to be used without disabling indexes. You could also consider including this information in release notes.

mzumsande commented:
> I do think it would be good if PR description described benefits of these changes

Done, thanks.

> You could also consider including this information in release notes.

Do you mean a sentence like "The `-reindex-chainstate` option can now be used without the necessity to disable custom indexes"? I'm not sure if the first change changes user experience enough to be part of the release notes.

mzumsande commented:
[53ff637 ](https://github.com/bitcoin/bitcoin/commit/53ff63799404e6828779c5e4f321acf256fe9018)to [2d0acc2](https://github.com/bitcoin/bitcoin/commit/2d0acc259836e5cb7f40579e76375420a942b86c): Rebased

ryanofsky reviewed (approved):
Code review ACK 2d0acc259836e5cb7f40579e76375420a942b86c. This is a simple, helpful PR. Would encourage others to review.

Only changes since last review are mentioning benefits of the change in the PR description (I added a few formatting and wording tweaks on top of this) and rebasing.

> Do you mean a sentence like "The -reindex-chainstate option can now be used without the necessity to disable custom indexes"? I'm not sure if the first change changes user experience enough to be part of the release notes.

Fair enough. I think it's a nice feature, but no need to mention it

mzumsande commented:
[2d0acc2 ](https://github.com/bitcoin/bitcoin/commit/2d0acc259836e5cb7f40579e76375420a942b86c)to [0d00d69](https://github.com/bitcoin/bitcoin/commit/0d00d6996c25d8b82412f022dfe246887bc135c4): rebased due to conflict with #25704

mzumsande commented:
0577bd6d1bcbbc808fd20d7a368a562799ceccc3 to 974140f9e721740f857b45d10d7dbab62fdbbe53: rebased due to conflict with #25781

furszy reviewed (commented):
Instead of the global flag that requires manual sets at different locations and the indexes threads active wait, what if we move the indexes threads start after the loading process? e.g. https://github.com/furszy/bitcoin-core/commit/1525e0ae360d10cbc8bd5d8f15bc27112254dbef.

It makes code shorter and more robust. Plus, it let us keep the pruning checks as well.

willcl-ark reviewed (approved):
ACK 974140f9e7

Can confirm via testing that this fixes the majority of #27558, although I was not able to reliably abort my node _during_ an `invalidateblock` call as OP did in that issue...

I also extracted the feature_coinstatsindex test from 5fafeec471 and checked that it did indeed fail without the corresponding changes to `BaseIndex::Init()`.

pinheadmz commented on src/index/base.cpp:
```diff
@@ -94,14 +95,22 @@ bool BaseIndex::Init()
         SetBestBlockIndex(nullptr);
     } else {
-        SetBestBlockIndex(m_chainstate->FindForkInGlobalIndex(locator));
+        // Setting the best block to the locator's top block. If it is not part of the
+        // best chain, we will rewind to the fork point during index sync
+        const CBlockIndex* locator_index{m_chainstate->m_blockman.LookupBlockIndex(locator.vHave.at(0))};
+        if (!locator_index) {
+            return InitError(strprintf(Untranslated("%s: best block of the index not found. Please rebuild the index."), GetName()));
+        }
+        SetBestBlockIndex(locator_index);
```
Another side effect of this PR is I don't think we use the `locator` at all anymore besides its tip hash ?!

mzumsande replied:
That's correct! I didn't want to change the db format though to not break compatibility.

pinheadmz commented:
> In first commit vector `front()` call is replaced by `at(0)` call to avoid undefined behavior in case null locator is loaded.

This happens again later too:

https://github.com/bitcoin/bitcoin/blob/974140f9e721740f857b45d10d7dbab62fdbbe53/src/index/base.cpp#L100

pinheadmz reviewed (approved):
ACK 974140f9e721740f857b45d10d7dbab62fdbbe53

code review and local testing. verified the tests fail without the patches. great bug catch on the rewinding muhash! I also like @furszy idea about dropping the global atomic bool for a rerranged init sequence. I'll be happy to re-review if you included that.

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 974140f9e721740f857b45d10d7dbab62fdbbe53
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmRdKuQACgkQ5+KYS2KJ
yTrWzRAAi1TzWndTKGoM6WwPvnXacA3OkluidZqIoa6zkeEnYzE7voAkRs2ZVmZp
T9FJ90ouWfkIfoaC1SewYGpJ2dykVSP41dmCgo0F06ceigiOAHTENDqJxkjFNL3e
+3sau/iPaGSOw8LU863kfABCI0KSqffnW9drQPySCt6lhCqSEhUJBpdBHHtaXS2z
/pF7nTUbvhV7hKbUKWB6bdLLkDO8w54yZXQkt3iNQDFWATFmutqxc7nhpdNe1fQW
vOqAOdw3ZdmPVNJB5H9nzmx46HDJemEr/w/8jy+6Hl6RAaxlhIs2a4gR64jLEeg6
8OD69S4XpCxu4FosZwoJyp7FmHL0ZINh4tR6bAPDO9Oevg5SFtsRP3zy01NI0PRW
wS3drLRG7PVkuFuwUHcExhPQyQIUoGnhLopfPjtngRTjfEuNUlUoYaZHzzO7Yjl6
eR6WG16vgtCHilnV1vypBX06U5gnq2Lu5VHoNTFuh9cveaIxXBKXtnJVLPhgaAmI
mMTfl+2A/2rvS3PgVXRBa+16e42h8y6rJB+6oHghhAD4Arz6sfn48K1blWk1Sp1j
NgNtREdWeqMmRVPTutn7Aoxa3/k7NeBC5qeipZdviSOqTusvuzt2mYZsIzVw7Aie
bwoX0quWu8tftAI4oEfGnRjXmJ9+XTJourAfZ9RDCJWzdYG6PXo=
=HWSL
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

mzumsande commented:
Thanks! I'll rebase and address furszy's comments next week!

ryanofsky commented:
> Thanks! I'll rebase and address furszy's comments next week!

I think it'd be good to just rebase this and merge it and not try to do the "move the indexes threads start after the loading process" change here. This PR is pretty simple, has had a good amount of review and testing, and I think that change would make it more complicated. furszy also implemented that change separately in #27607, and it should simplify both PRs to base that change on top of this one.

ryanofsky reviewed (approved):
Code review ACK 974140f9e721740f857b45d10d7dbab62fdbbe53. Confirmed this is just a clean rebase since my last review. This needs another rebase now, but after that I would like to merge it.

ryanofsky reviewed (approved):
Code review ACK 97844d9268b87b5d09b1091bfd0326ed18ce5b91. Just simple rebase since last review
